commands/auto/followtrajectory.py

```diff
@@ -0,0 +1,213 @@
+                    self.addRequirements(
+                        reqs
+                    )  # make sure we require everything our sub commands require
+
+        self.setControllers()
+
+        self.originTrajectory = trajectory
+        self.trajectory = trajectory
+
+        self.timer = Timer()
+
+        self.addRequirements([self.drive])
+        self.setName(__class__.__name__)
+
+    def setControllers(self) -> None:
+        self.xController = PIDController(
+            constants.kTrajectoryPositionPGain,
+            constants.kTrajectoryPositionIGain,
+            constants.kTrajectoryPositionDGain,
+        )
+        self.yController = PIDController(
+            constants.kTrajectoryPositionPGain,
+            constants.kTrajectoryPositionIGain,
+            constants.kTrajectoryPositionDGain,
+        )
+        self.thetaController = ProfiledPIDControllerRadians(
+            constants.kTrajectoryAnglePGain,
+            constants.kTrajectoryAngleIGain,
+            constants.kTrajectoryAngleDGain,
+            TrapezoidProfileRadians.Constraints(
+                constants.kMaxRotationAngularVelocity,
+                constants.kMaxRotationAngularAcceleration,
+            ),
+        )
+        self.thetaController.enableContinuousInput(-pi, pi)
+
+        self.controller = HolonomicDriveController(
+            self.xController, self.yController, self.thetaController
+        )
+
+    def initialize(self):
+        self.timer.reset()
+        self.timer.start()
+
+        self.unpassedMarkers = []
+        self.unpassedMarkers.extend(self.markers)
+
+        initialState = self.trajectory.getInitialState()
+        allianceRespectiveStartingPoint = self.getAllianceRespectivePoint(
+            initialState.pose, initialState.holonomicRotation
+        )
+        self.drive.resetOdometryAtPosition(
+            Pose2d(
+                allianceRespectiveStartingPoint[0].translation(),
+                allianceRespectiveStartingPoint[1],
+            ),
+        )
+
+        self.setControllers()
+
+        transformedStates = [
+            self.getAllianceRespectivePoint(state.pose, state.pose.rotation())
+            for state in self.originTrajectory.getStates()
+        ]
+
+        SmartDashboard.putNumberArray(
+            constants.kAutonomousPathKey,
+            functools.reduce(
+                operator.add,
+                [
+                    [state[0].X(), state[0].Y(), state[1].radians()]
+                    for state in transformedStates
+                ],
+                [],
+            ),
+        )
+        DataLogManager.log("begin trajectory")
+
+    def getAllianceRespectivePoint(
+        self, pose: Pose2d, holonomicRotation: Rotation2d
+    ) -> Tuple[Pose2d, Rotation2d]:
+        if DriverStation.getAlliance() == DriverStation.Alliance.kRed:
+            return (
+                Pose2d(
+                    constants.kFieldLength - pose.X(),
+                    pose.Y(),
```

Comment:
If we start the robot on the field, how does it have the right position? This would work in sim but idk about on field.

Reply:
this function you are looking at converts a pose and rotation into one that takes into account the translation required in blue vs red